tornado/routing.py

```diff
@@ -0,0 +1,335 @@
+import re
+from functools import partial
+
+from tornado import httputil
+from tornado.httpserver import _CallableAdapter
+from tornado.escape import url_escape, url_unescape, utf8
+from tornado.log import app_log
+from tornado.util import basestring_type, import_object, re_unescape, unicode_type
+
+
+class Router(httputil.HTTPServerConnectionDelegate):
+    def add_handlers(self, *args):
+        pass
+
+    def find_handler(self, request, **kwargs):
+        raise NotImplementedError()
+
+    def start_request(self, server_conn, request_conn):
+        return _RoutingDelegate(self, request_conn)
+
+
+class ReversibleRouter(Router):
+    def reverse_url(self, name, *args):
+        raise NotImplementedError()
+
+
+class _RoutingDelegate(httputil.HTTPMessageDelegate):
+    def __init__(self, router, request_conn):
+        self.connection = request_conn
+        self.delegate = None
+        self.router = router
+
+    def headers_received(self, start_line, headers):
+        request = httputil.HTTPServerRequest(
+            connection=self.connection, start_line=start_line,
+            headers=headers)
+
+        self.delegate = self.router.find_handler(request)
+        return self.delegate.headers_received(start_line, headers)
+
+    def data_received(self, chunk):
+        return self.delegate.data_received(chunk)
+
+    def finish(self):
+        self.delegate.finish()
+
+    def on_connection_close(self):
+        self.delegate.on_connection_close()
+
+
+class RuleRouter(Router):
+    def __init__(self, rules=None):
+        self.rules = []
+        if rules:
+            self.add_handlers(rules)
+
+    def add_handlers(self, handlers):
+        for rule in handlers:
+            if isinstance(rule, (tuple, list)):
+                assert len(rule) in (2, 3, 4)
+                if isinstance(rule[0], basestring_type):
+                    rule = URLSpec(*rule)
+                else:
+                    rule = Rule(*rule)
+
+            self.rules.append(self.process_rule(rule))
+
+    def process_rule(self, rule):
+        return rule
+
+    def find_handler(self, request, **kwargs):
+        for rule in self.rules:
+            target_params = rule.matcher(request)
+            if target_params is not None:
+                if rule.target_kwargs:
+                    target_params['target_kwargs'] = rule.target_kwargs
+
+                delegate = self.get_target_delegate(
+                    rule.target, request, **target_params)
+
+                if delegate is not None:
+                    return delegate
+
+        return None
+
+    def get_target_delegate(self, target, request, **target_params):
```

Comment:
A RequestHandler subclass mostly: ApplicationRouter (the one that knows about Application -> RequestHandler relation) overrides this method in order to provide a suitable delegate https://github.com/tornadoweb/tornado/pull/1806/files#diff-035ec61639a96ccea9c8c0e302cb243bR1749